tests/v3/connection_state_test.c

```diff
@@ -3387,25 +3387,120 @@ static int s_test_mqtt_connection_ping_no_fn(struct aws_allocator *allocator, vo
     ASSERT_SUCCESS(aws_mqtt_client_connection_connect(state_test_data->mqtt_connection, &connection_options));
     s_wait_for_connection_to_complete(state_test_data);
 
+    struct aws_byte_cursor pub_topic = aws_byte_cursor_from_c_str("/test/topic");
+    struct aws_byte_cursor payload_1 = aws_byte_cursor_from_c_str("Test Message 1");
+
+    /* Publish */
+    uint16_t packet_id = aws_mqtt_client_connection_publish(
+        state_test_data->mqtt_connection,
+        &pub_topic,
+        AWS_MQTT_QOS_AT_LEAST_ONCE,
+        false,
+        &payload_1,
+        s_on_op_complete,
+        state_test_data);
+    ASSERT_TRUE(packet_id > 0);
+
+    /* Wait 0.8 seconds */
+    aws_thread_current_sleep(800000000);
+
+    /* Subscribe */
+    packet_id = aws_mqtt_client_connection_subscribe(
+        state_test_data->mqtt_connection,
+        &pub_topic,
+        AWS_MQTT_QOS_AT_LEAST_ONCE,
+        s_on_publish_received,
+        state_test_data,
+        NULL,
+        s_on_suback,
+        state_test_data);
+    ASSERT_TRUE(packet_id > 0);
+
+    /* Wait 0.8 seconds */
+    aws_thread_current_sleep(800000000);
+
+    /* Resub */
+    uint16_t resub_packet_id =
+        aws_mqtt_resubscribe_existing_topics(state_test_data->mqtt_connection, s_on_multi_suback, state_test_data);
+    ASSERT_TRUE(resub_packet_id > 0);
+    /* Wait 0.8 seconds */
+    aws_thread_current_sleep(800000000);
+
+    /*Unsub*/
+    packet_id = aws_mqtt_client_connection_unsubscribe(
+        state_test_data->mqtt_connection, &pub_topic, s_on_op_complete, state_test_data);
+    ASSERT_TRUE(packet_id > 0);
+
+    /* Wait 0.8 seconds */
+    aws_thread_current_sleep(800000000);
+
+    /* Ensure the server got 0 PING packets */
+    ASSERT_INT_EQUALS(0, mqtt_mock_server_get_ping_count(state_test_data->mock_server));
+
+    ASSERT_SUCCESS(
+        aws_mqtt_client_connection_disconnect(state_test_data->mqtt_connection, s_on_disconnect_fn, state_test_data));
+    s_wait_for_disconnect_to_complete(state_test_data);
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE_FIXTURE(
+    mqtt_connection_ping_no,
+    s_setup_mqtt_server_fn,
+    s_test_mqtt_connection_ping_no_fn,
+    s_clean_up_mqtt_server_fn,
+    &test_data)
+
+/**
+ * Makes a CONNECT, with 1 second keep alive ping interval, disable the server auto ack so that we never received ack
+ * back. Send a qos1 publish roughly every second for 4 seconds. As we never received the ACK, we should send a total of
+ * 4 ping
```

Comment:
This is alarming.  Given a keep aliveof 1 second and a ping timeout of 100ms we should be seeing multiple disconnections which are going to disrupt the logical simplicity of the test.